File: server/agent/src/agents/recommender.js
const llm = require('../llm/connector');
const db = require('../db/database');
const { RECOMMENDER_PROMPT, getLanguageInstruction } = require('../prompts/system-prompts');
const { formatPlaceForLLM, filterByInterests } = require('../utils/helpers');
const { rankPlaces } = require('./scoring');
const { enforceDiversity } = require('./diversity');

async function recommend(params) {
    const {
        query = '', cities = [], interests = [], limit = 8,
        language = 'en', group_type = 'family', budget_level = 'medium'
    } = params;

    let candidates = [];

    if (query) {
        candidates = db.searchPlaces(query);
    }

    if (candidates.length < limit * 3) {
        const cityList = cities.length > 0 ? cities : db.getCities();
        for (const city of cityList) {
            const cityPlaces = db.getPlacesByCity(city);
            candidates.push(...cityPlaces);
        }
        candidates = [...new Map(candidates.map(p => [p.id, p])).values()];
    }

    if (interests.length > 0) {
        const filtered = filterByInterests(candidates, interests);
        if (filtered.length >= 3) candidates = filtered;
    }

    const intent = { interests, cities, group_type, budget_level };
    candidates = rankPlaces(candidates, intent);

    candidates = enforceDiversity(candidates, {
        maxPerCategory: 3,
        cities,
        minPerCity: cities.length > 0 ? 2 : 0,
        totalNeeded: Math.min(25, candidates.length),
        surfaceHiddenGems: true
    });

    console.log(`[Recommender] ${candidates.length} diversified candidates (top: ${candidates.slice(0, 3).map(p => p.name).join(', ')})`);

    const topCandidates = candidates.slice(0, Math.min(25, candidates.length));
    const formatted = topCandidates.map(formatPlaceForLLM);

    const userPrompt = JSON.stringify({
        user_query: query,
        preferences: { interests, cities, group_type, budget_level },
        available_places: formatted
    });

    const result = await llm.generateJSON(RECOMMENDER_PROMPT + getLanguageInstruction(language), userPrompt);

    if (!result.success) {
        
        return {
            success: true,
            recommendations: topCandidates.slice(0, limit).map(p => ({
                place_id: p.id,
                name: p.name,
                city: p.city,
                category: p.category,
                why: p.highlight
                    ? `⭐ Must-visit highlight in ${p.city} — ranked by relevance & diversity`
                    : `Recommended ${p.category} in ${p.city} (score: ${p.score || 'N/A'})`,
                best_time: p.best_time_to_visit || 'Morning',
                duration: `${p.estimated_visit_duration} mins`,
                crowd_level: p.crowd_level,
                insider_tip: p.highlight ? '⭐ Must-visit highlight!' : 'Worth a visit'
            })),
            summary: `Top ${limit} places ranked by relevance, diversity, and preferences`,
            source: 'fallback'
        };
    }

    return {
        success: true,
        ...result.data,
        source: result.source
    };
}

module.exports = { recommend };

File: server_py/agent/agents/recommender.py
"""
RecommenderAgent — Place recommendation specialist.
"""

import json
from typing import Optional

from agent.agents.core.agent_base import AgentBase
from agent.prompts.system_prompts import RECOMMENDER_PROMPT
from agent.db import database as db
from agent.utils.helpers import format_place_for_llm
from agent.agents.scoring import rank_places
from agent.agents.diversity import enforce_diversity
from agent.llm import connector as llm


class RecommenderAgent(AgentBase):
    """Agent specialized in personalized place recommendations."""

    def __init__(self):
        super().__init__(
            name="RecommenderAgent",
            system_prompt=RECOMMENDER_PROMPT,
            tool_names=[
                "search_places", "get_places_by_city", "get_places_by_interest",
                "get_highlights", "get_category_distribution", "get_ranked_places",
            ],
            max_iterations=4,
            memory_limit=5,
        )

    async def recommend(self, params: dict) -> dict:
        """Generate place recommendations.

        Args:
            params: { cities?, interests?, group_type?, budget_level?, query? }

        Returns:
            { type, text, recommendations?, source }
        """
        cities = params.get("cities", [])
        interests = params.get("interests", [])
        query = params.get("query", "")

        # Get places from DB
        if cities:
            places = db.get_places_by_multiple_cities(cities)
        else:
            places = db.get_all_places()

        # Rank and diversify
        intent = {
            "interests": interests,
            "group_type": params.get("group_type", "family"),
            "budget_level": params.get("budget_level", "medium"),
            "cities": cities,
        }
        ranked = rank_places(places, intent)
        diversified = enforce_diversity(ranked, {
            "cities": cities,
            "totalNeeded": 15,
            "minPerCity": 3,
            "maxPerCategory": 4,
        })

        slim_places = [format_place_for_llm(p) for p in diversified]

        # Build prompt
        prompt_addition = f"""
User query: {query or "Recommend places to visit"}
User interests: {', '.join(interests) if interests else 'General'}
Cities: {', '.join(cities) if cities else 'All Braj cities'}
Group type: {params.get('group_type', 'not specified')}
Budget: {params.get('budget_level', 'not specified')}

Available places (ranked by relevance):
{json.dumps(slim_places[:12], ensure_ascii=False)}
"""

        result = await llm.generate_json(
            self.system_prompt,
            prompt_addition,
        )

        if result.get("success") and result.get("data"):
            data = result["data"]
            return {
                "type": "recommendations",
                "text": data.get("summary", ""),
                "recommendations": data.get("recommendations", []),
                "source": self.name,
            }

        # Fallback: return from the ReAct loop
        run_result = await self.run(
            f"Recommend places based on these preferences: {json.dumps(params)}",
            params,
        )
        return {
            "type": "text",
            "text": run_result.get("text", ""),
            "source": self.name,
        }
